### PR TITLE
Make EntryImpl recyclable to improve gc at broker

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation

Broker creates instance of [EntryImpl](https://github.com/yahoo/pulsar/blob/master/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java) when it receives a published-msg and delivers to consumer. Broker also keeps it into cache to deliver it to already caught-up consumers. However, `EntryImpl` is not recyclable which impacts young gc as broker can create this object multiple times for each message.

### Modifications

- Made `EntryImpl` recyclable.
- Made `PositionImpl` object's life-cycle independent to `EntryImpl` so, recycling of `EntryImpl` will not impact any other existing flow. It also gives a scope to make `PositionImpl` recyclable in future or will be helpful while replacing existing CacheImpl with other implementation.

### Result

It reduces young gc pause and time at broker.

### CODE REVIEW

**Comment by rdhabalia** [comment]
after running system-test with and without change into perf environment, we can see young-gc improvements from avg 450msec to 150ms

**Before change:**
ML-Cache ................................................................................................GC
screenshot before-change

**After change**
ML-Cache ................................................................................................GC
screenshot after-change

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/Entry.java [hunk 1]
```diff
@@ -48,7 +50,18 @@
...
     /**
-     * Release the resources allocated for this entry
+     * Increments ref-count of entry data
+     */
+    void retain();
+    
+    /**
+     * Recycles entry instance 
+     */
+    void recycle();
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/Entry.java]: Since there is already a ref-counting semantic on the class, can we take advantage of it and automatically recycle when ref-count goes to 0?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java [hunk 2]
```diff
@@ -22,36 +22,60 @@
...
+        }
+    };
+    
+    private final Handle recyclerHandle;
+    private PositionImpl position;
+    private ByteBuf data;
+
+    public static EntryImpl create(LedgerEntry ledgerEntry) {
+        EntryImpl entry = RECYCLER.get();
+        entry.position = new PositionImpl(ledgerEntry.getLedgerId(), ledgerEntry.getEntryId());
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java]: Even better we could reuse the `PositionImpl` instance as well. We just need to make `ledgerId` and `entryId` non final on that class.

Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java]: Regarding the life-cycle of the `PositionImpl`, we could return a new `PositionImpl` instance in `getPosition()` itself. The advantage would be that the object that is being retained for longer time (while in the cache) is going to get reused

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java [hunk 3]
```diff
@@ -174,7 +175,8 @@ public String consumerName() {
                         .build();
 
                 ByteBuf metadataAndPayload = entry.getDataBuffer();
-
+                // increment ref-count of data and release at the end of process: so, we can get chance to call entry.release
+                entry.retain();
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java]: The semantic here is kind of weird, it's kind of mix-matching ref-counting and recycler

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java [hunk 4]
```diff
@@ -298,13 +300,16 @@ public synchronized void readEntriesComplete(List<Entry> entries, Object ctx) {
             int messagesForC = Math.min(Math.min(entriesToDispatch, c.getAvailablePermits()), MaxRoundRobinBatchSize);
 
             if (messagesForC > 0) {
-                int msgSent = c.sendMessages(entries.subList(start, start + messagesForC)).getRight();
-
+                
+                // remove positions first from replay list first : sendMessages recycles entries
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java]: This changes seems to be not related with this PR, right?

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java (hunk 3 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java]: Actually, earlier when dispatcher gives entries to [Consumer](https://github.com/yahoo/pulsar/pull/318/files/81e154d1597389f60ca3f1678449d408cb32131e#diff-95b9bf24e8d791dbf6026b1a64984900), consumer retrieves `data(byteBuf)` from it and puts bytebuf into netty-channel which decrements the ref-count by 1 (and ByteByf will be recycled if ref-count=0).

Now, as we made `EntryImpl` recyclable and it gets recycled when we call `Entry.release()` and if data ref-count =0, we recycle the `EntryImpl`. So, we `retain()` which helps later on to call [entry.release()](https://github.com/rdhabalia/pulsar/blob/81e154d1597389f60ca3f1678449d408cb32131e/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java#L201) and if entry's data ref-count is 0 then we recycle Entry and data both.

I could have avoided retain and could have recycled Entry in the [end](https://github.com/rdhabalia/pulsar/blob/81e154d1597389f60ca3f1678449d408cb32131e/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java#L201) But caller(eg. dispatcher) may not want to recycle and if caller has retain data for later use then we may break this contract. So, this scope retain and in the last try to release the entry (and it gets recycled if data ref-count=0).

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java [hunk 5]
```diff
@@ -141,7 +141,11 @@ public void safeRun() {
         ManagedLedgerImpl.TOTAL_SIZE_UPDATER.addAndGet(ml, dataLength);
         if (ml.hasActiveCursors()) {
             // Avoid caching entries if no cursor has been created
-            ml.entryCache.insert(new EntryImpl(ledger.getId(), entryId, data));
+            EntryImpl entry = EntryImpl.create(ledger.getId(), entryId, data);
+            // EntryCache.insert: duplicates entry by allocating new entry and data. so, recycle entry after calling
+            // insert
+            ml.entryCache.insert(entry);
+            entry.recycle();
```
Comment (rdhabalia) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java]: As we discussed, `ml.entryCache.insert(entry)` creates new `EntryImpl` internally and allocates new data to store into cache. So, life-cycle of this entry is finished after calling `cache.insert(entry)` so, recycling of this entry will not affect cache data because cache is already having duplicate-created entry.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/Entry.java (hunk 1 above)
Comment (rdhabalia) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/Entry.java]: we follow the same semantic (same as ByteBuf -> release() will recycle if ref-count=0). However, we provided this method to support usecase where one wants to use EntryImpl temp to just pass the data(byteBuf) and not disturbing life-cycle of that `data` and recycle that EntryImpl immediately. eg. [OpAddEntry](https://github.com/yahoo/pulsar/pull/318/files/81e154d1597389f60ca3f1678449d408cb32131e#diff-772b4979830bd173bc8fa5a1d52a7588) or [Consumer](https://github.com/rdhabalia/pulsar/blob/81e154d1597389f60ca3f1678449d408cb32131e/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java#L201)

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java (hunk 4 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java]: it is related to this PR. in this change we remove entries from replay list before we send those entries to consumer. Because it's consumer's responsibility to release entry's data and recycle it. 
Now, to remove positions from replay list after sending to consumer, we have to read positions from already recycled entires which gives NPE.
So, we remove first from replay list and then send to consumer.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java (hunk 2 above)
Comment (rdhabalia) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java]: yes, make sense. I will make the change.

Comment (rdhabalia) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java]: @merlimat made the change.

**Comment by saandrews** [comment]
@merlimat Do you have any more concern? We are trying to get this merged today if there is no blocker.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java (hunk 3 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java]: yes, that's correct. So, if we remove `release()` then we have to call `entry.data.release()` and `entry.recycle()` explicitly. So, we have two options:

1. Every time make two explicit call
```
entry.getDataBuffer().release();
entry.recycle();
```

2. rename `release()` with `releaseAndRecycle()` which clarifies purpose => release data and recycle entry. and caller doesn't have to call multiple things.

Do, you think we 2nd option will solve this concern or we should go with the 1st one.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java [hunk 6]
```diff
@@ -80,24 +105,37 @@ public int getLength() {
...
+        }
     }
 
-    public int refCnt() {
-        return data.refCnt();
+    @Override
+    public void recycle() {
+        this.data = null;
+        // recycle position if it is recyclable object
+        this.position.recycle();
```
Comment (rdhabalia) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java]: >Recyclying on the PositionImpl is currently disabled (and actually it might be good to remove that dead-code in this PR as well)

I have already cleaned up dead-code and enabled the recycling on [PositionImpl](https://github.com/rdhabalia/pulsar/blob/f3199810d6997e026a4dd0dc0bbf6593e0d327a2/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/PositionImpl.java)

>we can reuse the same position all the time, we just need to reset the ledgerId and entryId on it.

Don't you think recycle would be better as it can protect to mess up things and it will surely return null on already recycled entryImpl.getPosition() rather reset value.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java [hunk 7]
```diff
@@ -141,7 +141,11 @@ public void safeRun() {
         ManagedLedgerImpl.TOTAL_SIZE_UPDATER.addAndGet(ml, dataLength);
         if (ml.hasActiveCursors()) {
             // Avoid caching entries if no cursor has been created
-            ml.entryCache.insert(new EntryImpl(ledger.getId(), entryId, data));
+            EntryImpl entry = EntryImpl.create(ledger.getId(), entryId, data);
```
Comment (rdhabalia) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java]: yes, should we try to address change which can prevent duplicate data in to different PR after we merge this one.??

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java (hunk 5 above)
Comment (rdhabalia) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java]: you mean we should just insert same entry object to cache without copying it?
```
        if (ml.hasActiveCursors()) {
            EntryImpl entry = EntryImpl.create(ledger.getId(), entryId, data);
            ml.entryCache.insert(entry); // It should not duplicate and insert the same entry
        } else {
          data.release();
        }
```
If yes then I don't see any issue but then do you know any reason behind it having in existing [code](https://github.com/yahoo/pulsar/blob/master/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java#L144)..??

Comment (rdhabalia) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java]: I think as per documentation the reason is buffer coming from netty is in 64kb chunk so, to avoid storing unnecessary memory for smaller entry < 64kb, we copy and store it. So, I think we should change definition of insert as `insert(ByteBuf)` as we don't use passed `EntryImpl` object in method at all. any thought?

**Comment by merlimat** [comment]
@rdhabalia @saandrews Added 1 commit on top of this PR. Take a look at https://github.com/merlimat/pulsar/commit/08de7c7c47fdb242074b08ddbb81a3fae8b2f72b 

Few things: 
 1. Use proper ref-count sematic for EntryImpl. EntryImpl owns a ref of the buffer (so, while we have an `EntryImpl`, the buffer is alive.
 2. Removed recycling on PositionImpl, this was long due since it's impossible to achieve on the current premises. Use ledgerId/entryId in entryImpl so that is more compact

**Comment by rdhabalia** [comment]
@merlimat Looks good to me. By introducing refCounted `EntryImpl`, we init it with refCount=1 so, it will be recycled immediately once we call `release` on it. So, we may have issue if that same entry is being used anywhere after releasing it.
I will cherry-pick this change into this PR and I am verifying if we have such usecase so, we can address it after this change.

**Comment by merlimat** [comment]
> So, we may have issue if that same entry is being used anywhere after releasing it.

Yes, take it with a grain of salt.. it would be good to have it go through a regression as well

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryCacheImpl.java [hunk 8]
```diff
@@ -118,12 +118,14 @@ public boolean insert(EntryImpl entry) {
             entryBuf.readerIndex(readerIdx);
         }
 
-        if (entries.put(entry.getPosition(), new EntryImpl(entry.getPosition(), cachedData))) {
+        PositionImpl position = entry.getPosition();
+        EntryImpl cacheEntry = EntryImpl.create(position, cachedData);
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryCacheImpl.java]: I forgot to release `cachedData` here. Since `cacheEntry` retains the buffer, we need to do `cachedData.release()` just after this line

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/PositionImpl.java [hunk 9]
```diff
@@ -34,49 +34,31 @@
     private long ledgerId;
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/PositionImpl.java]: forgot to put back `final` for `ledgerId` and `entryId`
